File: src/main/scala/xenit/care4alf/module/RestErrorHandling.scala
package xenit.care4alf.module

import com.github.dynamicextensionsalfresco.webscripts.annotations.ExceptionHandler
import org.springframework.extensions.webscripts.WebScriptResponse
import com.typesafe.scalalogging.slf4j.Logging

/**
 * @author Laurent Van der Linden
 */
trait RestErrorHandling extends Logging {
    @ExceptionHandler(Array(classOf[IllegalArgumentException]))
    def handleIllegalArgument(exception: Exception, response: WebScriptResponse) {
        logger.error("Illegal argument", exception)
        response.setStatus(404)
        response.getWriter.write(exception.getMessage)
    }
}

File: src/main/kotlin/eu/xenit/care4alf/web/RestErrorHandling.kt
package eu.xenit.care4alf.web

import org.springframework.extensions.webscripts.WebScriptResponse
import org.slf4j.Logger
import com.github.dynamicextensionsalfresco.webscripts.annotations.ExceptionHandler
import org.json.JSONWriter
import javax.servlet.http.HttpServletResponse

/**
 * @author Laurent Van der Linden
 */
public interface RestErrorHandling {
    var logger: Logger

    ExceptionHandler(Exception::class)
    fun handleIllegalArgument(exception: Exception, response: WebScriptResponse) {
        logger.error("Controller failure", exception)

        var cause: Throwable = exception
        while (cause.getCause() != null) {
            cause = cause.getCause()!!
        }

        if (cause is IllegalArgumentException) {
            response.setStatus(HttpServletResponse.SC_BAD_REQUEST)
        } else {
            response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR)
        }
        JSONWriter(response.getWriter()).`object`()
            .key("message").value(cause.getMessage())
        .endObject()
    }
}